### PR TITLE
check usingFifo() for getFramesRead/Written()

### LINKED ISSUE #153: Audio*StreamOpenSLES should check usingFifo() in getFramesRead/Written
If the stream is in blocking read/write mode then getFramesRead() and getFramesWritten() should return the number of frames related to the Oboe FIFO, not the internal mixer.

if (usingFifo()) return AudioStreamBuffered::getFramesRead();
else ...

### PR DESCRIPTION
**Pull request by philburk**
For better frame counting in OpenSL ES.
When a FIFO was being used, getFramesRead() was returning
the server position. It should return the FIFO position.

Fixes #153

### CODE REVIEW

File: src/aaudio/AudioStreamAAudio.cpp [hunk 1]
```diff
@@ -221,14 +221,17 @@ Result AudioStreamAAudio::open() {
...
-Result AudioStreamAAudio::close()
-{
+Result AudioStreamAAudio::close() {
     // The main reason we have this mutex if to prevent a collision between a call
     // by the application to stop a stream at the same time that an onError callback
     // is being executed because of a disconnect. The close will delete the stream,
     // which could otherwise cause the requestStop() to crash.
     std::lock_guard<std::mutex> lock(mLock);
 
+    // Update frame counter variables to avoid retrograde motion.
```
Comment (dturner) [inline src/aaudio/AudioStreamAAudio.cpp]: You're updating something by calling `get` methods which feels weird. Could you use a setter instead or even `incrementFramesRead(0)` at a push?

File: src/opensles/AudioStreamBuffered.cpp [hunk 2]
```diff
@@ -40,19 +42,20 @@ void AudioStreamBuffered::allocateFifo() {
...
         }
-        mFifoBuffer.reset(new FifoBuffer(getBytesPerFrame(), capacity));
+        // TODO consider using std::make_unique if we require c++14
+        mFifoBuffer.reset( new FifoBuffer(getBytesPerFrame(), capacity));
     }
 }
 
 int64_t AudioStreamBuffered::getFramesWritten() {
-    if (usingFIFO()) {
+    if (mFifoBuffer) {
```
Comment (dturner) [inline src/opensles/AudioStreamBuffered.cpp]: Would prefer explicit check `mFifoBuffer != nullptr` to make it clear that `mFifoBuffer` isn't a bool.

Comment (dturner) [reply src/opensles/AudioStreamBuffered.cpp]: Ah right, I see later on you changed `mFifoBuffer` to a `unique_ptr` which has a bool operator. As you were :)

File: src/opensles/AudioStreamOpenSLES.cpp [hunk 3]
```diff
@@ -216,6 +216,10 @@ Result AudioStreamOpenSLES::close() {
     if (mState == StreamState::Closed){
         return Result::ErrorClosed;
     } else {
+        // Update frame counter variables so they can be read after close.
+        getFramesWritten();
```
Comment (dturner) [inline src/opensles/AudioStreamOpenSLES.cpp]: Again would prefer not to update something using get methods.

File: src/opensles/AudioStreamBuffered.cpp (hunk 2 above)
Comment (philburk) [reply src/opensles/AudioStreamBuffered.cpp]: I'm trying to be more C++ish. ;-)

File: src/aaudio/AudioStreamAAudio.cpp (hunk 1 above)
Comment (philburk) [reply src/aaudio/AudioStreamAAudio.cpp]: Right. That does feel wierd. I will add an update method.

File: include/oboe/AudioStream.h [hunk 4]
```diff
@@ -395,6 +395,16 @@ class AudioStream : public AudioStreamBase {
 
     AudioFormat mNativeFormat = AudioFormat::Invalid;
 
+    /**
+     * Update mFramesWritten.
```
Comment (dturner) [inline include/oboe/AudioStream.h]: Does this need to be a public method? It feels like this should only ever be called by the stream class (or subclass) to update its own frame counter from the underlying stream, which isn't something a client should be concerned with. If you agree, suggest changing to `protected`.

File: include/oboe/AudioStream.h [hunk 5]
```diff
@@ -395,6 +395,16 @@ class AudioStream : public AudioStreamBase {
...
+     */
+    virtual void updateFramesWritten() = 0;
+
+    /**
+     * Update mFramesRead.
+     */
+    virtual void updateFramesRead() = 0;
```
Comment (dturner) [inline include/oboe/AudioStream.h]: As above. Also might be worth moving the 2 member variables below into a protected section as well. The more we can reduce the API surface the less likely we will be to break it.

File: include/oboe/AudioStream.h (hunk 4 above)
Comment (philburk) [reply include/oboe/AudioStream.h]: It is already "protected:".

File: include/oboe/AudioStream.h (hunk 5 above)
Comment (philburk) [reply include/oboe/AudioStream.h]: It is already "protected:".

**Pull request merged**